### PR TITLE
introduce bootstrapper for coap-gw cert init

### PR DESCRIPTION
**Pull request by ondrejtomcik**

### CODE REVIEW

File: coap-gateway/bootstrapper/go.mod [hunk 1]
```diff
@@ -0,0 +1,13 @@
+module github.com/plgd-dev/cloud/coap-gateway/bootstrapper
```
Comment (jkralik) [inline coap-gateway/bootstrapper/go.mod]: pls remove `go.mod` file - we use only per repository so `github.com/plgd-dev/cloud/go.mod`

File: coap-gateway/bootstrapper/go.sum [hunk 2]
```diff
@@ -0,0 +1,1418 @@
+cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Comment (jkralik) [inline coap-gateway/bootstrapper/go.sum]: pls remove go.sum file - we use only per repository so github.com/plgd-dev/cloud/go.sum

File: coap-gateway/bootstrapper/main.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+			if bs.isCertificateReady(certRequest) {
+				bs.createSecret(pkeyBlock, certRequest.Status.Certificate, certRequest.Status.CA)
+				bs.cleanupCertificateRequest()
+				close(informerStopCh)
+			}
+		},
+	})
+	go informer.Run(informerStopCh)
+
+	_, err := bs.certManagerClient.CertmanagerV1().CertificateRequests(bs.namespace).Create(context.TODO(), &certRequest, metaV1.CreateOptions{})
```
Comment (jkralik) [inline coap-gateway/bootstrapper/main.go]: use proper `ctx` instead of `context.TODO()`

Comment (jkralik) [reply coap-gateway/bootstrapper/main.go]:
```
ctx, cancel := contest.WithCancel(context.Background())
defer cancel()
...
```

File: coap-gateway/bootstrapper/main.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+	}
+	bs.namespace = strings.TrimSuffix(string(namespace), "\n")
+}
+
+func (c *Configuration) Validate() error {
+	if c.CloudID == "" {
+		return fmt.Errorf("cloudID is required")
+	}
+	if _, err := uuid.Parse(c.CloudID); err != nil {
+		return fmt.Errorf("cloudID('%v')", c.CloudID)
```
Comment (jkralik) [inline coap-gateway/bootstrapper/main.go]: pls pass error message to wrapped error

File: coap-gateway/bootstrapper/main.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+	bs.cleanupCertificateRequest()
+	factory := cmInformers.NewSharedInformerFactory(bs.certManagerClient, 0)
+	informer := factory.Certmanager().V1().CertificateRequests().Informer()
+	informerStopCh := make(chan struct{})
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			certRequest := obj.(*cmV1.CertificateRequest)
+			if bs.isCertificateReady(certRequest) {
+				bs.createSecret(pkeyBlock, certRequest.Status.Certificate, certRequest.Status.CA)
```
Comment (jkralik) [inline coap-gateway/bootstrapper/main.go:80]: pass `ctx` as first argument of `bs.createSecret = `bs.createSecret(ctx,....)`

File: coap-gateway/bootstrapper/main.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+		if latestStatusIndex >= 0 {
+			latestStatus := certRequest.Status.Conditions[latestStatusIndex]
+			log.Printf("status of certificate request %v updated: %v", bs.configuration.SecretName, latestStatus)
+			return latestStatus.Type == cmV1.CertificateRequestConditionReady && latestStatus.Status == cmMetaV1.ConditionTrue
+		}
+	}
+	return false
+}
+
+func (bs *Bootstrapper) createSecret(pkeyBlock, cert, ca []byte) {
```
Comment (jkralik) [inline coap-gateway/bootstrapper/main.go:194]: add `ctx context.Context` as first argument

File: coap-gateway/bootstrapper/main.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+			Name: bs.configuration.SecretName,
+		},
+		Data: map[string][]byte{
+			"tls.crt": cert,
+			"tls.key": pkeyBlock,
+			"ca.crt":  ca,
+		},
+	}
+
+	_, err = bs.kubernetesClient.CoreV1().Secrets(bs.namespace).Get(context.TODO(), bs.configuration.SecretName, metaV1.GetOptions{})
```
Comment (jkralik) [inline coap-gateway/bootstrapper/main.go]: use ctx

File: coap-gateway/bootstrapper/main.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+	if errors.IsNotFound(err) {
+		_, err = bs.kubernetesClient.CoreV1().Secrets(bs.namespace).Create(context.TODO(), &secretSpec, metaV1.CreateOptions{})
```
Comment (jkralik) [inline coap-gateway/bootstrapper/main.go]: use ctx

File: coap-gateway/bootstrapper/main.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+		if err != nil {
+			log.Fatalf("unable to create secret: %v", err)
+		}
+		log.Printf("secret %v created", bs.configuration.SecretName)
+	} else if err == nil {
+		_, err = bs.kubernetesClient.CoreV1().Secrets(bs.namespace).Update(context.TODO(), &secretSpec, metaV1.UpdateOptions{})
```
Comment (jkralik) [inline coap-gateway/bootstrapper/main.go]: use ctx

**Review by jkralik — COMMENTED** [review]
Pls use proper `ctx` instead of `context.TODO()`

File: coap-gateway/bootstrapper/go.mod (hunk 1 above)
Comment (ondrejtomcik) [reply coap-gateway/bootstrapper/go.mod]: done

File: coap-gateway/bootstrapper/go.sum (hunk 2 above)
Comment (ondrejtomcik) [reply coap-gateway/bootstrapper/go.sum]: done

File: coap-gateway/bootstrapper/main.go (hunk 3 above)
Comment (ondrejtomcik) [reply coap-gateway/bootstrapper/main.go]: done

File: coap-gateway/bootstrapper/main.go (hunk 5 above)
Comment (ondrejtomcik) [reply coap-gateway/bootstrapper/main.go:80]: done

File: coap-gateway/bootstrapper/main.go (hunk 6 above)
Comment (ondrejtomcik) [reply coap-gateway/bootstrapper/main.go:194]: done

File: coap-gateway/bootstrapper/main.go (hunk 7 above)
Comment (ondrejtomcik) [reply coap-gateway/bootstrapper/main.go]: done

File: coap-gateway/bootstrapper/main.go (hunk 8 above)
Comment (ondrejtomcik) [reply coap-gateway/bootstrapper/main.go]: done

File: coap-gateway/bootstrapper/main.go (hunk 9 above)
Comment (ondrejtomcik) [reply coap-gateway/bootstrapper/main.go]: done

File: coap-gateway/bootstrapper/main.go (hunk 4 above)
Comment (ondrejtomcik) [reply coap-gateway/bootstrapper/main.go]: done

File: go.mod [hunk 10]
```diff
@@ -14,17 +15,19 @@ require (
...
 	github.com/jessevdk/go-flags v1.5.0
+	github.com/jetstack/cert-manager v1.3.1
 	github.com/jtacoma/uritemplates v1.0.0
 	github.com/karrick/tparse/v2 v2.8.2
 	github.com/kelseyhightower/envconfig v1.4.0
 	github.com/lestrrat-go/jwx v1.2.0
 	github.com/nats-io/nats.go v1.11.0
 	github.com/panjf2000/ants/v2 v2.4.5
 	github.com/patrickmn/go-cache v2.1.0+incompatible
+	github.com/plgd-dev/cqrs v0.0.0-20200819120550-a6063edff616 // indirect
```
Comment (jkralik) [inline go.mod:26]: pls remove dep

File: coap-gateway/bootstrapper/main.go [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+		},
+		Data: map[string][]byte{
+			bs.configuration.CertFileName: cert,
+			bs.configuration.KeyFileName:  pkeyBlock,
+			bs.configuration.CAFileName:   ca,
+		},
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
```
Comment (jkralik) [inline coap-gateway/bootstrapper/main.go:217]: this is wird ... call this in caller function and pass ctx as argument

File: coap-gateway/bootstrapper/main.go (hunk 5 above)
Comment (jkralik) [reply coap-gateway/bootstrapper/main.go:80]: Not resovlved it is much better when you pass ctx as argument.

File: coap-gateway/bootstrapper/main.go (hunk 11 above)
Comment (jkralik) [reply coap-gateway/bootstrapper/main.go:217]: why constant 20sec ? why it is not configurable ?

**Pull request closed without merging**